[ManagerValidator] Public test harness API

foundrytom opened the pull request:
Refactors the manager validator to expose the limited public API as described in #177.

feltech commented on tests/openassetio/test/managerValidator/test_harness.py:
```diff
@@ -0,0 +1,222 @@
+        # pylint: disable=protected-access
+        assert case._session == mock_session
+        assert case._fixtures == a_fixture_dict
+        assert case._manager == mock_session.currentManager()
```
[consider][minor] 
```suggestion
        assert case._manager == mock_session.currentManager.return_value
```
Otherwise you're adding to the call count of the mock.  Not a problem in this test, just a minor nitpick.

foundrytom replied:
Fixed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/cfed515eebb8263036a24fe1588dcf20a5aa08b7).

feltech commented on tests/openassetio/test/managerValidator/conftest.py:
```diff
@@ -19,18 +19,33 @@
-def mock_host_interface():
-    return mock.create_autospec(
-        validatorHarness.ValidatorHarnessHostInterface, instance=True, spec_set=True)
+def test_resources_dir():
```
[consider] Here and below, ideally don't prefix fixtures with `test_`, it's a bit confusing.

foundrytom replied:
Oh man, yes it so is! 🤣  How about [this](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/69c5b5abacfb9aa5e730ce687edcd002b0753317)?

feltech commented on python/openassetio/test/managerValidator/_implementation.py:
```diff
@@ -0,0 +1,151 @@
+
+class _ValidatorHarness:
+    """
+    Facade wrapping the unittest runner.
```
[consider][minor] My bad, it's not really a "facade", or at least not only facade, more like an "adapter", or maybe we should be hand-wavy and just say "wrapper".

foundrytom replied:
Hows [this](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/1761c2208e3871c94b14e5d11c1634c15083280d)?

feltech commented on tests/openassetio/test/managerValidator/resources/executeSuiteTests.py:
```diff
@@ -0,0 +1,82 @@
+#
```
[consider][minor]  It would be nice to name this file and its fixtures to better resemble the other files in this directory.  Maybe `customTestSuite` / `fixtures_custom`, or `metaTestSuite` :grin: , or maybe similar to before but suffixed to align more closely with other files, `executeSuiteTestSuite` / `fixtures_executeSuite`... as you often say, naming things is hard!

foundrytom replied:
How about:
![image](https://user-images.githubusercontent.com/12052093/149373251-1ce6e07b-a76b-4cb4-8abe-d073cb74056b.png)

feltech replied:
Yeah much nicer!

foundrytom replied:
Implemented [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/1bda909a486d09b3f420260ff7f5238095369a3c), along side the module renames we discussed, [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/621e30af1f197889ee74e0e38e0f9495fbaf258c) and [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/42707bff24c3b959bbc153ea59ecc6ee0ec60fd6).

feltech commented on python/openassetio/test/managerValidator/_implementation.py:
```diff
@@ -0,0 +1,151 @@
+            for testCaseName in testCaseNames)
+        return loadedSuite
+
+    def setFixtures(self, fixtures):
```
[question] Curious about (from commit message) "[a]llows harness to be re-used with different modules", is this something we envision?  Or was it mainly to more closely align fixture and suite lifetimes?

Ideally we would have a test for swapping out different fixtures with the same `_ValidatorTestLoader`, but currently that's not possible via the public interface so not a big deal.

foundrytom replied:
I was just to try and align the lifetimes. In that, the implementation allowed the harness to be re-run with different modules (which we don't currently do), but those would all have to share fixtures, as there was no way to change the fixtures once the loader was constructed and supplied when the harness was constructed.  Happy to drop this commit though.

feltech replied:
No it's fine, sounds reasonable to me.

foundrytom commented:
@feltech As per our chat, I've squashed a bunch of those commits down. If you're happy with these intermediate fixups, I'll merge those too, then do the remaining tidying/renaming.

feltech commented:
> @feltech As per our chat, I've squashed a bunch of those commits down. If you're happy with these intermediate fixups, I'll merge those too, then do the remaining tidying/renaming.

Not sure if you'd already planned this, but the commit message for 0649ebb could do with a bit of tidying.

foundrytom commented:
> > @feltech As per our chat, I've squashed a bunch of those commits down. If you're happy with these intermediate fixups, I'll merge those too, then do the remaining tidying/renaming.
> 
> Not sure if you'd already planned this, but the commit message for [0649ebb](https://github.com/TheFoundryVisionmongers/OpenAssetIO/commit/0649ebb5aa8128ba310183c61ac0f5e8282c7a40) could do with a bit of tidying.

Yeah, need to sort all the messages out. Just fixed those names, and the other renames. If you're happy with those, I'll tidy the commit messages up.

feltech commented on python/openassetio/test/manager/_implementation.py:
```diff
@@ -0,0 +1,151 @@
+    """
+    # pylint: disable=too-few-public-methods
+
+    _kValidatorHarnessProgramName = "manager"
```
[consider] Similar to other comment, this is for argparse help messages, so that e.g.
```
$ OPENASSETIO_PLUGIN_PATH=resources/examples/manager/SampleAssetManager/python python -m openassetio.test.manager -f resources/examples/manager/SampleAssetManager/test/fixtures.py --some-unsupported-switch
usage: manager [-h] [-v] [-q] [--locals] [-f] [-c] [-b] [-k TESTNAMEPATTERNS] [tests [tests ...]]
manager: error: unrecognized arguments: --some-unsupported-switch
```
so we probably want something a bit more descriptive than `manager`.

foundrytom replied:
Fixed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/6699dbda1241566581b328ceda68e93ec22d2134).

feltech commented on tests/openassetio/test/manager/resources/fixtures_cliFail.py:
```diff
@@ -30,7 +30,7 @@
     "Test_displayName": {
         "test_matches_fixture": {
-            "displayName": "managerValidator Stub Manager"
+            "displayName": "manager Stub Manager"
```
[consider][minor] Maybe just "Stub Manager"?

foundrytom replied:
Yeah good idea, fixed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/174fbae7371264271307239306d60f8eacead9d8).

feltech commented on tests/openassetio/test/manager/resources/fixtures_cliPass.py:
```diff
@@ -14,13 +14,13 @@
 #
 """
-Manager test harness test case fixtures for managerValidator test
+Manager test harness test case fixtures for manager test
```
[consider][minor] A bit confusing now it says "manager test", perhaps "manager test harness tests"

foundrytom replied:
Reworded [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/72aebad69a5ebe77d2a97ab61116d5402d7749a2).

feltech commented on tests/openassetio/test/manager/resources/suite_executeSuiteTests.py:
```diff
@@ -0,0 +1,82 @@
+    def test_when_called_then_session_has_expected_host(self):
+        self.assertEqual(
+            self._session.host().identifier(),
+            "org.openassetio.test.manager"
```
[consider] Perhaps for clarity it would be worth naming the stub host identifier more explicitly, e.g. 
```suggestion
            "org.openassetio.test.manager.stubHost"
```

foundrytom replied:
Improved [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/0bcc7baac69a49057d52b19d50d88ec782154eed). I didn't use 'stub' in the end - as technically, it felt like it was a real host - ie, the test harness is a totally legit host in its own right.

foundrytom commented on python/openassetio/test/manager/__main__.py:
```diff
@@ -0,0 +1,93 @@
+
+
+cmdline = argparse.ArgumentParser(
+    prog="manager",
```
Good call, sorry had missed all of these side effects as I was distracted by the prospect of lunch. Bad form. Fixed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/184/commits/ea18c7565162a1074abd6c304c2318e28ed2037d).

feltech reviewed (approved):
LGTM squash and merge when ready
